File: transactions/src/main/java/com/example/config/SecurityConfig.java
package com.example.config;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.web.SecurityFilterChain;

@Configuration
public class SecurityConfig {

    private final String authServerUrl;

    public SecurityConfig(@Value("${auth.server.url}")String authServerUrl) {
        this.authServerUrl = authServerUrl;
    }

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
        http
                .authorizeHttpRequests(authorize -> authorize
                        .requestMatchers("/api/accounts**").authenticated()
                        .anyRequest().permitAll())
                .oauth2ResourceServer(oauth2 -> oauth2
                        .jwt(jwt -> jwt
                                .jwkSetUri(authServerUrl + "/oauth2/jwks")));
        return http.build();
    }

}

File: transactions/src/main/kotlin/com/example/config/SecurityConfig.kt
package com.example.config

import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.web.server.ServerHttpSecurity
import org.springframework.security.web.server.SecurityWebFilterChain


@Configuration
class SecurityConfig(
        @Value("\${auth.server.url}") private val authServerUrl: String
) {

    @Bean
    fun springSecurityFilterChain(http: ServerHttpSecurity): SecurityWebFilterChain {
        return http
                .authorizeExchange { exchanges ->
                    exchanges
                            .pathMatchers("/api/accounts/**").authenticated()
                            .anyExchange().permitAll()
                }
                .oauth2ResourceServer { oauth2 ->
                    oauth2.jwt { jwt ->
                        jwt.jwkSetUri("$authServerUrl/oauth2/jwks")
                    }
                }
                .build()
    }
}
